Add Danielle's new-and-improved navigation

Paul-Hebert opened the pull request:
This PR adds Danielle's new navigation which has been optimized to handle large lists of patterns and prototypes without overwhelming users, as well as work better on phones. (It also looks real nice!)

dromo77 committed:
Add large screen navigation

dromo77 committed:
Add an aside on large screens

dromo77 committed:
Add jump links on pattern pages

dromo77 committed:
Hide overflow for demo purposes

dromo77 committed:
Update caret icon

dromo77 committed:
Update submenu styles on small screens

dromo77 committed:
Update styles

dromo77 committed:
Add border styles

dromo77 committed:
Add transition styles

dromo77 committed:
Add shadow styles

dromo77 committed:
Update card styles

dromo77 committed:
Update navigation structure

dromo77 committed:
Update aside

dromo77 committed:
Add semi sticky navigation

dromo77 committed:
Change position of sidebar on large screens

dromo77 committed:
Update comment

dromo77 committed:
Update layout

dromo77 committed:
Change cursor for buttons and accordions

dromo77 committed:
Increase touch/click area of nav links

dromo77 committed:
Increase color contrast

dromo77 committed:
Fix nav overflow issue

dromo77 committed:
Update sticky nav function

dromo77 committed:
Add headsup.js

dromo77 committed:
Merge branch 'master' into ui-exploration-7

dromo77 committed:
Install headsup and update nav styles

dromo77 committed:
Add pattern and prototype examples

dromo77 committed:
Update structure of navigation

dromo77 committed:
Adjust card styles

dromo77 committed:
Update nav layout on small screens

dromo77 committed:
Vertically align nav items on large screens

dromo77 committed:
Adjust sidebar position

Paul-Hebert committed:
Remove dummy content

Paul-Hebert commented on src/_includes/layouts/pattern-library.hbs:
```diff
@@ -4,43 +4,85 @@ styles:
+    </header>
+    <nav class="EP_nav">
+      <ul class="EP_nav__menu">
+        <li class="EP_nav__category">
```
These `li`s are repeated and should be a reusable partial

Paul-Hebert commented on src/_includes/layouts/pattern-library.hbs:
```diff
@@ -4,43 +4,85 @@ styles:
+    <nav class="EP_nav">
+      <ul class="EP_nav__menu">
+        <li class="EP_nav__category">
+          <a class="EP_nav__submenu-toggle" aria-expanded="false" role="button">
```
Would it be preferable to use an actual button over `role="button"`?

dromo77 replied:
We could, if we do use a button I think we should consider styling the focus state.

Paul-Hebert replied:
I think it would be better to use a button. The [first rule of ARIA](https://www.deque.com/blog/top-5-rules-of-aria/) is to only use ARIA when there isn't a native HTML element that can fill the role. So in this case it would be better to use a native button than `aria-role="button"`

We should definitely add a focus state as well! Good call! We may also need one for the Menu button if we don't have one. These focus styles could potentially be broken out into a separate issue if that's easier.

dromo77 replied:
>I think it would be better to use a button.

👍 I can update the markup to use a button element. 

>These focus styles could potentially be broken out into a separate issue if that's easier.

I'm leaning towards creating a separate issue so that we can wrap up this work.

Paul-Hebert replied:
Cool, sounds good. Thanks :)

Paul-Hebert commented on src/_includes/layouts/pattern-library.hbs:
```diff
@@ -4,43 +4,85 @@ styles:
-      <div class="EP_nav__category">
-        <h3>
-          <a href="/prototypes">
+      <button class="EP_nav__menu-toggle EP_button" aria-expanded="false">Menu</button>
```
I forget how `aria-expanded` targets the expanded collapsed area. Can you point to an ID or something for some screen readers?

Paul-Hebert commented on src/_includes/layouts/pattern-library.hbs:
```diff
@@ -4,43 +4,85 @@ styles:
+        <li class="EP_nav__category">
+          <a class="EP_nav__submenu-toggle" aria-expanded="false" role="button">
+              Prototypes
+            <svg xmlns="http://www.w3.org/2000/svg" width="20" height="10" viewBox="0 0 20 12.35" class="toggle-icon" aria-hidden="true" focusable="false">
```
It seems odd that the width/height ratio on this SVG doesn't match the viewbox ratio.

Paul-Hebert commented on src/_includes/layouts/pattern-sidebar.hbs:
```diff
@@ -0,0 +1,20 @@
+---
+
+<div class="EP_layout__sidebar">
+  <ul class="EP_sidebar__menu">
```
I would expect there to be an `EP_sidebar` element for the BEM naming to be consistent. Maybe this needs to be broken into separate components more clearly?

Paul-Hebert commented on src/_includes/layouts/prototype-sidebar.hbs:
```diff
@@ -0,0 +1,20 @@
+---
```
I think we should be able to use the same partial for both sidebars.

Paul-Hebert commented on src/scripts/eleventy-patterns/navigation.js:
```diff
@@ -0,0 +1,41 @@
+
+// Toggle menu on small screens
+function initMenuToggle() {
+  const menuToggle = document.querySelector('.EP_nav__menu-toggle');
```
We should use `js-` prefixed class selectors here.

Paul-Hebert replied:
(And throughout this file)

Paul-Hebert commented:
I can still sometimes get stuck in loops of the header showing and hiding

Paul-Hebert commented on src/scripts/eleventy-patterns/navigation.js:
```diff
@@ -0,0 +1,41 @@
+  const menu = document.querySelector('.EP_nav__menu');
+
+  menuToggle.addEventListener('click', () => {
+    let expanded = menuToggle.getAttribute('aria-expanded') === 'true' || false;
```
I'm a little confused by this `|| false` pattern here.

Paul-Hebert commented on src/scripts/eleventy-patterns/navigation.js:
```diff
@@ -0,0 +1,41 @@
+  menuToggle.addEventListener('click', () => {
+    let expanded = menuToggle.getAttribute('aria-expanded') === 'true' || false;
+    menu.classList.toggle('is-open');
+    menuToggle.setAttribute('aria-expanded', !expanded);
```
I'm confused why we're setting `aria-expanded` to the opposite of `expanded`

Paul-Hebert replied:
Oh, I get it... I'm going to add a comment explaining this.

Paul-Hebert commented on src/scripts/eleventy-patterns/navigation.js:
```diff
@@ -0,0 +1,41 @@
+  submenuToggle.forEach(toggle => {
+    toggle.addEventListener('click', event => {
+      event.preventDefault();
+      let target = toggle.nextElementSibling;
```
I think target could be a const

Paul-Hebert commented on src/scripts/eleventy-patterns/navigation.js:
```diff
@@ -0,0 +1,41 @@
+      let expanded = toggle.getAttribute('aria-expanded') === 'true' || false;
+      const toggleIcon = toggle.querySelector('.toggle-icon');
+      target.classList.toggle('is-open');
+      toggleIcon.classList.toggle('is-open');
```
It would be nice if we could just set the class on target, and have the icon's CSS change based on the target's class

Paul-Hebert commented on src/scripts/eleventy-patterns/navigation.js:
```diff
@@ -0,0 +1,41 @@
+    debounce: 100,
+    hiddenHeaderClass: 'EP-nav-hidden'
+  });
+});
```
I wonder if this file should be imported into the main pattern lib JS bundle (I haven't viewed all the files yet, so maybe it is!)

Paul-Hebert replied:
I think this is probably fine as-is for now. If we add more pattern lib JS we may rethink this.

Paul-Hebert commented on src/styles/eleventy-patterns/config/borders.scss:
```diff
@@ -0,0 +1,14 @@
+  --border-size-thick: 4px;
+  --border-size-medium: 2px;
+  --border-size-thin: 1px;
+}
```
I wonder if we need all these... if we do they should be EP prefixed

Paul-Hebert commented on src/styles/eleventy-patterns/config/shadows.scss:
```diff
@@ -0,0 +1,17 @@
+    var(--shadow-blur)
+    var(--shadow-color);
+
+}
```
These should be EP prefixed

Paul-Hebert commented on src/styles/eleventy-patterns/config/transitions.scss:
```diff
@@ -0,0 +1,47 @@
+/* prettier-ignore */
```
Do we need this for the whole file?

Paul-Hebert commented on src/styles/eleventy-patterns/config/transitions.scss:
```diff
@@ -0,0 +1,47 @@
+  --transition-glacial: var(--ease-normal-out) var(--duration-glacial);
+  --transition-geological: var(--ease-normal-out) var(--duration-geological);
+  --transition-geological-delay: var(--transition-geological) var(--duration-prompt);
+  --transition-default: var(--transition-prompt);
```
I'm not sure this repo needs this many transitions. May be overkill.

This should all be EP prefixed

Paul-Hebert commented on src/styles/eleventy-patterns/index.scss:
```diff
@@ -13,5 +13,24 @@
   height: 100%;
 }
+
+body.no-scroll {
```
no-scroll should be EP prefixed

Paul-Hebert commented on src/styles/eleventy-patterns/index.scss:
```diff
@@ -13,5 +13,24 @@
 
 body,
 html {
+  color: var(--EP_color-text);
```
This needs to be EP namespaced somehow

Paul-Hebert commented on src/styles/eleventy-patterns/index.scss:
```diff
@@ -13,5 +13,24 @@
+
+a {
+  color: var(--EP_color-blue);
+  text-decoration: none;
```
I'm not sure how I feel about removing the underline. All of this should be EP namespaced.

Paul-Hebert commented on src/styles/eleventy-patterns/patterns/button.scss:
```diff
@@ -0,0 +1,18 @@
+    background: var(--EP_color-dark-grey);
+    color: var(--EP_color-white);
+  }
+}
```
I worry that the contrast between the button and its background may be too low. Maybe a border would help?

Paul-Hebert commented on src/styles/eleventy-patterns/patterns/nav.scss:
```diff
@@ -1,9 +1,145 @@
+}
+
+.EP-nav-hidden .EP_header-container {
+  box-shadow: 0 0 0 rgba(67, 90, 111, 0);
```
Is this meant to remove the shadow? I think we could just use `none`

Paul-Hebert replied:
Or we could move the header further off screen

Paul-Hebert commented on src/styles/eleventy-patterns/patterns/nav.scss:
```diff
@@ -1,9 +1,145 @@
+  width: 100%;
+  z-index: 1;
+
+  @media (min-width: 800px) {
```
We should use `em`s for breakpoints. And maybe store them somewhere

dromo77 replied:
I updated all media queries to use `em`s, but it isn't stored anywhere.

Paul-Hebert replied:
Perfect, thanks. Would you mind opening a Github issue to standardize breakpoints in custom properties or SCSS vars?

Paul-Hebert commented on src/styles/eleventy-patterns/config/sizes.scss:
```diff
@@ -0,0 +1,5 @@
+:root {
+  --EP_nav-height: 80px;
```
Should this be rem/em?

Paul-Hebert commented on src/styles/eleventy-patterns/patterns/nav.scss:
```diff
@@ -1,9 +1,145 @@
+    display: flex;
+    height: var(--EP_nav-height);
+    justify-content: space-between;
+    padding: 0 1.5em;
```
It would be nice to make more of these commons sizes (1.5, 1.15, etc.) into custom props. We could also use a modular scale. Maybe that's a follow-up issue?

dromo77 replied:
I think adding some common sizes is a good idea. I'm on the fence about using a modular scale. It could be nice to already have a sizing system set up. However, I'm unsure if people will want to use the same sizing that is being used in the project for the pattern library. What do you think?

Paul-Hebert replied:
>However, I'm unsure if people will want to use the same sizing that is being used in the project for the pattern library.

I think we should do our best to keep our boilerplate CSS separate from the CSS our end users will be writing. The way we've generally handled this is with namespacing. If we create custom sizes, or implement a modular scale, we'd want to namespace that as well.

I think this is likely out of the scope of this PR though 🙂

dromo77 replied:
That makes sense, I can add an issue for this.

Paul-Hebert replied:
Perfect, thanks

Paul-Hebert commented on src/styles/eleventy-patterns/patterns/nav.scss:
```diff
@@ -1,9 +1,145 @@
+  text-decoration: none;
+}
+
+.toggle-icon {
```
This should be EP namespaced

dromo77 committed:
Use EP prefixed class names and remove css bloat

dromo77 committed:
Use "EP_js" to prefix javascript hooks

dromo77 committed:
Update caret icon

dromo77 committed:
Remove box shadow from header when hidden

dromo77 committed:
Update header height to use rem

dromo77 committed:
Update nav hidden class

dromo77 committed:
Add border to button

dromo77 committed:
Update breakpoint to use em units

dromo77 committed:
Update naming structure for main content

dromo77 committed:
Use buttons to toggle accordion sections

Paul-Hebert committed:
Remove unnecessary `|| false`

Paul-Hebert committed:
Add explanatory comments about expanded toggle

Paul-Hebert committed:
Use `const` over `let` and minor code reorg

Paul-Hebert committed:
Rework submenu logic

Simplify so only one element needs a class toggled.

Target all elements with selectors instead of nextSibling to be more stable.

Paul-Hebert committed:
Better namespace global pattern lib styles

Paul-Hebert committed:
Make a partial for the repeated sidebar

Paul-Hebert committed:
Use partial for repeated menu markup

Paul-Hebert committed:
Fix header jank

As the body padding was toggled you could get in situations where the header showed and hid itself over and over.

Paul-Hebert committed:
Move mobile menu next to its trigger

So aria-expanded maps to the correct element.

As part of this work I also made some changes to the header to remove extraneous elements and make things more semantic

Paul-Hebert committed:
Fix gap under mobile menu when scrolling

Paul-Hebert committed:
Lock scrolling when mobile menu expanded

Paul-Hebert committed:
Remove anchor tag styles

Since they're not namespaced

Paul-Hebert committed:
Revert accidental commit

dromo77 commented:
This is looking great @Paul-Hebert, thanks for your help tackling these changes! The only difference I noticed is the primary links on large screens are underlined. I wanted to check if this was a conscious decision. I know we decided to keep the text-decoration visible in other places.

Paul-Hebert committed:
Fix nav primary link underline regression

Paul-Hebert commented:
Good catch @dromo77 ! That wasn't intentional. It likely happened when I removed the `a` styling we talked about. Should be fixed now!

dromo77 commented:
Thanks, this is looking great!
